---
title: "ZA short read MAGs"
author: "Fiona Tamburini"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: paper
    code_folding: hide
    toc: true
    toc_float: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```

```{r}
library(cowplot)
library(ggplot2)
library(ggpubr)
library(here)
library(kableExtra)
library(tidyverse)

## params ----
global_pal <- c("#E3211C", "#F89897", "#6A3D9A", "#CAB2D6", "#1F78B4", "#A5CEE3")
za_pal <- global_pal[3:4]

## read input_final data ----
## sample metadata
za_meta <- readRDS(here("rds/za_meta.rds"))

## genomesearch output
gsearch <- read.table(here("input_final/uhgg_ANI_compare/perident_table_uhgg_formatted.tsv"),
                      sep = "\t", header = T,
                      col.names = c("Sample", "Bin", "gsearch.ANI",
                                    "gsearch.Phylum", "gsearch.Species"))

## fastani output
fastani <- read.table(here("input_final/uhgg_ANI_compare/fastani_top.txt"),
                      sep = "\t", header = F,
                      col.names = c("Sample", "Bin","UHGG_genome", "fastani.ANI",
                                    "fragment_mappings", "query_fragments", "AF"))

# take top match per mag
# some mags missing due to low identity
fastani_top <- fastani %>%
  arrange(Sample, Bin, -fastani.ANI) %>%
  group_by(Sample, Bin) %>%
  slice(1)

## gtdbtk classifications
gtdbtk_bac <- read.table(here("input_final/mags/gtdbtk.bac120.summary.tsv"),
                     sep = "\t", header = T)
gtdbtk_arc <- read.table(here("input_final/mags/gtdbtk.ar122.summary.tsv"),
                     sep = "\t", header = T)
gtdbtk <- rbind(gtdbtk_bac, gtdbtk_arc)
gtdbtk <- gtdbtk %>%
  separate(user_genome, into = c("Sample", "Bin"), sep = "_", extra = "merge")

## dereplication lists
# za 95% ANI
drep_za_95 <- read.table(here("input_final/mags/za_drep95_reps.txt"), sep = "\t",
                      header = F, col.names = c("Sample", "Bin"))

# za 95% ANI
drep_za_99 <- read.table(here("input_final/mags/za_drep99_reps.txt"), sep = "\t",
                      header = F, col.names = c("Sample", "Bin"))

# za plus uhgg reps, 95% ANI
drep_za_uhgg <- read.table(here("input_final/mags/za_uhggreps_drep95_reps.txt"),
                           sep = "\t", header = F, col.names = c("Sample", "Bin"))

## binning dastool pipeline output
mags <- read.table(here("input_final/mags/binning_table_all_simple.tsv"),
                   sep = "\t", header = T, comment.char = "")

## format mag info ----
# add genomesearch and fastani comparisons
# only consider MQ and higher
mags <- mags %>%
  filter(Bin != "unbinned") %>%
  mutate(Quality = ifelse(Completeness > 90 & Contamination < 5 & tRNA >= 18 & rna.16S > 0 & rna.23S > 0 & rna.5S > 0,
                          "High-quality",
                          ifelse(Completeness >= 50 & Contamination <10, "Medium-quality", "Low-quality")),
         Quality = factor(Quality, levels = c("Low-quality", "Medium-quality", "High-quality"))) %>%
  left_join(za_meta, by = c("Sample" = "sample")) %>%
  left_join(gsearch, by = c("Sample", "Bin")) %>%
  left_join(fastani_top, by = c("Sample", "Bin")) %>%
  left_join(gtdbtk[, c("Sample", "Bin", "classification")], by = c("Sample", "Bin")) %>%
  mutate(za_rep95 = paste(Sample, Bin) %in% paste(drep_za_95$Sample, drep_za_95$Bin),
         za_rep99 = paste(Sample, Bin) %in% paste(drep_za_99$Sample, drep_za_99$Bin),
         uhgg_rep = paste(Sample, Bin) %in% paste(drep_za_uhgg$Sample, drep_za_uhgg$Bin))

mags_mqplus <- mags %>%
  filter(Quality != "Low-quality")

# mags table
mags %>%
  count(Quality) %>%
  kable() %>%
  kable_styling(full_width = F)

# save to supplementary table
labels <- read.table(here("input_final/za_labels.tsv"), sep = "\t", header = T)

mags_supp <- mags %>%
  left_join(labels[, c("sample", "label")], by = c("Sample" = "sample")) %>%
  filter(za_rep99) %>%
  select(label, Bin, site, best_species, classification, Completeness, Contamination,
         Strain.heterogeneity, Size.Mb, N50, Genes, tRNA, starts_with("rna"),
         za_rep95, uhgg_rep) %>%
  mutate(N50 = N50 / 1000)

names(mags_supp) <- c("Sample", "Bin", "Site", "NCBI species", "GTDB taxonomy",
                      "Completeness", "Contamination", "Strain heterogeneity",
                      "Size (Mb)", "N50 (Kb)", "# genes", "# tRNA",
                      "# 16S rRNA", "# 23S rRNA", "# 5S rRNA",
                      "dRep 95% sANI rep (cohort)",
                      "dRep 95% sANI rep (cohort plus UHGG species reps)")

write.table(mags_supp, here("final_tables/table_S10_drep_mags.txt"), sep = "\t",
            quote = F, row.names = F)
```

## Genome search ANI

MQ or better
```{r}
## ANI plots ----
## gsearch
ggplot(mags_mqplus, aes(site, gsearch.ANI, fill = site)) + 
  geom_jitter(position = position_jitterdodge(jitter.width = 0.7), alpha = 0.75, color = "darkgray") +
  geom_boxplot(outlier.shape = NA, alpha = 0.75) +
  # ylim(c(0, 100)) +
  ylab("Average Amino Acid Identity to UHGG") + 
  scale_fill_manual(values = za_pal) +
  theme_cowplot() +
  theme(legend.position = "none") +
  background_grid(major = "y")

ggsave(here("plots_v2/mags/gsearch_MAGs.png"), width = 5, height = 5)
```

## FastANI

MQ or better
```{r}
## fastANI
m <- mags_mqplus %>%
  mutate(site_abbrev = ifelse(site == "Soweto", "SWT", "BBR"))
a <- ggplot(m, aes(site_abbrev, fastani.ANI, fill = site_abbrev)) + 
  geom_jitter(position = position_jitterdodge(jitter.width = 0.7), alpha = 0.75, color = "darkgray") +
  geom_boxplot(outlier.shape = NA, alpha = 0.75) +
  # ylim(c(0, 100)) +
  labs(x = "", 
       y = "FastANI to UHGG") + 
  scale_fill_manual(values = za_pal) +
  theme_cowplot() +
  theme(legend.position = "none") +
  background_grid(major = "y") +
  geom_hline(yintercept = 95, linetype = "dashed")
a
ggsave(here("plots_v2/mags/fastANI_MAGs.png"), width = 5, height = 5)
```

## Comparison of genome search and fastANI
```{r}
## fastANI/genomesearch correlation
ggplot(mags_mqplus, aes(gsearch.ANI, fastani.ANI, color = site, size = Quality)) +
  geom_point(alpha = 0.75) +
  scale_size_manual(values = c(2, 3)) +
  scale_color_manual(values = za_pal) +
  theme_cowplot() +
  background_grid()

# top 95%
m <- mags_mqplus %>%
  filter(gsearch.ANI > 95,
         fastani.ANI > 95)

ggplot(m, aes(gsearch.ANI, fastani.ANI, color = site, size = Quality)) +
  geom_point(alpha = 0.75) +
  scale_size_manual(values = c(2, 3)) +
  scale_color_manual(values = za_pal) +
  theme_cowplot() +
  background_grid() +
  scale_x_log10() +
  scale_y_log10() +
  ylim(c(95, 100)) +
  xlim(c(95, 100)) +
  geom_abline(slope = 1, intercept = 0) +
  labs(color = "Site") +
  coord_fixed()

ggsave(here("plots_v2/mags/compare_genomesearch_fastANI.png"), width = 7, height = 5)
```

## Novel MAGs
```{r}
## novel mags ----
## what are the mags with low identity to uhgg?
novel_mags <- mags_mqplus %>%
  filter(is.na(fastani.ANI) | fastani.ANI < 95) %>%
  separate(classification, into = c("D", "P", "C", "O", "F", "G", "S"),
           sep = ";", remove = F)

mag_tbl <- novel_mags %>%
  select(Sample, Bin, bin.quality.call, Completeness, Contamination, classification)
DT::datatable(mag_tbl)
```

`r nrow(novel_mags)` novel MAGs with <95% identity to UHGG

### By phylum
```{r}
# by phylum
novel_mags %>%
  group_by(P) %>%
  tally() %>%
  arrange(-n) %>%
  kable() %>%
  kable_styling(full_width = F)
```

### By class
```{r}
# by class
novel_mags %>%
  group_by(C) %>%
  tally() %>%
  arrange(-n) %>%
  kable() %>%
  kable_styling(full_width = F)
```

## Representatives when dereplicated with UHGG species reps
```{r}
# za reps
# novel_mags %>%
#   filter(za_rep)

# za plus uhgg reps
novel_mags %>%
  filter(uhgg_rep) %>%
  select(Sample, Bin, bin.quality.call, Completeness, Contamination,
         classification, UHGG_genome, fastani.ANI) %>%
  DT::datatable(colnames = c("UHGG best match" = "UHGG_genome",
                             "UHGG best match fastANI" = "fastani.ANI"))
```

```{r, include = F}
b <- novel_mags %>%
  group_by(P) %>%
  mutate(P = gsub("p__", "", P)) %>%
  tally() %>%
  arrange(-n) %>%
  ggtexttable(rows = NULL, cols = c("GTDB phylum", "No. MAGs"),
              theme = ttheme("classic", base_size = 9,
                             padding = unit(c(4, 4), "mm")))

mgs <- mags_mqplus %>%
  arrange(-Completeness, Contamination)

c <- ggplot(mgs, aes(Completeness, Contamination, color = Quality, size = Quality)) +
  geom_point(alpha = 0.75) +
  scale_size_manual(values = c(1, 2)) +
  theme_cowplot() +
  background_grid() +
  theme(legend.position = "top",
        legend.justification = "center")

plot_grid(c,
          plot_grid(a, b, rel_widths = c(0.55, 0.45), labels = c("B", "C")),
          ncol = 1,
          rel_heights = c(0.55, 0.45),
          labels = c("A", ""))
ggsave(here("plots_v2/mags/novel_MAG_figure.png"), width = 5, height = 8)

```